### PR TITLE
Feature/permission events

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
- Adds Permission events and a list of events to the colony.
- Adds a way to add fakePlayers or players through the list to the colony.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java [hunk 1]
```diff
@@ -239,6 +239,70 @@ public void messageOnServerThread(final AddPlayer message, final EntityPlayerMP
         }
     }
 
+    /**
+     * Message class for adding a player or fakePlayer to a permission set.
+     */
+    public static class AddPlayerOrFakePlayer extends AbstractMessage<AddPlayerOrFakePlayer, IMessage>
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java:245]: MAJOR Reduce this class from 44 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 75 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 67 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L230](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L230): Remove this call from a constructor to the overridable "setupRsDataStore" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1367](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1367): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1437](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1437): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1468](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1468): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1480](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1480): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1583](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1583): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L48](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L48): class "WindowTownHall" has 38 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L823](https://github.com/Minecolonies/minecolonies/blob/365c1f2f2f46e18382b92d0bd74c544232a60c5b/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L823): Reduce this anonymous class number of lines from 41 to at most 40, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — COMMENTED** [review]
I saw you made it so that in PermissionEvent it is possible that the UUID is null. Did you check for that when Rendering and storing?

**Comment by Raycoms** [comment]
Yeah

**Comment by marchermans** [comment]
Okey

**Comment by marchermans** [comment]
Will pull this when I get home

**Review by marchermans — APPROVED** [review]

**Pull request merged**